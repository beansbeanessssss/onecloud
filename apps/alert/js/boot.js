define({

	 /**
 	 * allow to setup the application
 	 *
 	 * @return promise
	 * @param self will feed back 'this' object
 	 */

	setup: () => {
		var p = new Promise((r, d) => {
			requirejs([OC.appJS('alert', 'alert.bundle')], () => {
				r({
					id: 'alert',
					start : true
				});
			});
		});
		return p;
	},

	boot: () => { return false }
});
