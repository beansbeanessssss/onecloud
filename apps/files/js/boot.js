define({

	 	/**
		 * allow to setup the application
		 *
		 * @return promise
		 * @param self will feed back 'this' object
		 */

	 	setup: () => {
	 		var my = self.info;
	 		var p = new Promise((resolve, defer) => {
	 			OC.registerNavItem('files', {
	 				name: 'Files',
	 				iconMaterial: 'folder',
	 				route: '/'
	 			});
				resolve( { id : 'files' } );
	 		});
	 		return p;
	 	},


		/**
		 * Start the application by mounting it to the respective DOM element
		 *
		 * @return promise
		 */

		boot: (container) => {
			var p = new Promise((resolve, defer) => {
				requirejs([OC.appJS('files', 'files.bundle')], (app) => {
					app.$mount(container);
					app.$once('mounted', resolve( { start: true } ) );
				});
			});
			return p;
		}
});
