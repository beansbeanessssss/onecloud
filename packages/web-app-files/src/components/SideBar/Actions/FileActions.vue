<template>
  <oc-list id="oc-files-actions-sidebar" class="oc-mt-s">
    <action-menu-item
      v-for="(action, index) in actions"
      :key="`action-${index}`"
      :action="action"
      :items="resources"
      class="oc-py-xs"
    />
  </oc-list>
</template>

<script>
import { mapGetters } from 'vuex'
import ActionMenuItem from '../../ActionMenuItem.vue'

import FileActions from '../../../mixins/fileActions'
import { useIsFilesAppActive } from '../../../composables/useIsFilesAppActive'

export default {
  name: 'FileActions',
  components: { ActionMenuItem },
  mixins: [FileActions],
  setup() {
    return {
      isFilesAppActive: useIsFilesAppActive()
    }
  },
  title: ($gettext) => {
    return $gettext('Actions')
  },
  computed: {
    ...mapGetters('Files', ['highlightedFile']),

    resources() {
      return [this.highlightedFile]
    },

    actions() {
      return this.$_fileActions_getAllAvailableActions(this.resources)
    }
  }
}
</script>

<style lang="scss">
#oc-files-actions-sidebar {
  > li a,
  > li a:hover {
    color: var(--oc-color-swatch-passive-default);
    display: inline-flex;
    gap: 10px;
    vertical-align: top;
  }
}
</style>
