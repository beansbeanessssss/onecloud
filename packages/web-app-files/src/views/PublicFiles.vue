<template>
  <div class="oc-flex">
    <keyboard-actions :paginated-resources="paginatedResources" />
    <files-view-wrapper>
      <app-bar
        :has-bulk-actions="true"
        :breadcrumbs="breadcrumbs"
        :breadcrumbs-context-actions-items="[currentFolder]"
        :show-actions-on-selection="true"
        :side-bar-open="sideBarOpen"
      >
        <template #actions="{ limitedScreenSpace }">
          <create-and-upload :limited-screen-space="limitedScreenSpace" />
        </template>
      </app-bar>
      <app-loading-spinner v-if="areResourcesLoading" />
      <template v-else>
        <not-found-message v-if="folderNotFound" class="files-not-found oc-height-1-1" />
        <no-content-message
          v-else-if="isEmpty"
          id="files-public-list-empty"
          class="files-empty"
          icon="folder"
        >
          <template #message>
            <span v-translate>There are no resources in this folder</span>
          </template>
          <template v-if="currentFolder.canCreate()" #callToAction>
            <span v-translate data-testid="public-files-call-to-action">
              Drag files and folders here or use the "New" or "Upload" buttons to add files
            </span>
          </template>
        </no-content-message>
        <resource-table
          v-else
          id="files-public-files-table"
          v-model="selectedResourcesIds"
          class="files-table"
          :class="{ 'files-table-squashed': sideBarOpen }"
          :fields-displayed="['name', 'size', 'mdate']"
          :are-thumbnails-displayed="displayThumbnails"
          :resources="paginatedResources"
          :target-route="targetRoute"
          :header-position="fileListHeaderY"
          :sort-by="sortBy"
          :sort-dir="sortDir"
          :drag-drop="true"
          @fileDropped="fileDropped"
          @fileClick="$_fileActions_triggerDefaultAction"
          @rowMounted="rowMounted"
          @sort="handleSort"
        >
          <template #contextMenu="{ resource }">
            <context-actions v-if="isResourceInSelection(resource)" :items="selectedResources" />
          </template>
          <template #footer>
            <pagination :pages="paginationPages" :current-page="paginationPage" />
            <list-info
              v-if="paginatedResources.length > 0"
              class="oc-width-1-1 oc-my-s"
              :files="totalFilesCount.files"
              :folders="totalFilesCount.folders"
              :size="totalFilesSize"
            />
          </template>
        </resource-table>
      </template>
    </files-view-wrapper>
    <side-bar :open="sideBarOpen" :active-panel="sideBarActivePanel" />
  </div>
</template>

<script lang="ts">
import { mapGetters, mapActions, mapMutations } from 'vuex'
import ResourceTable from '../components/FilesList/ResourceTable.vue'
import { useResourcesViewDefaults } from '../composables'

import MixinAccessibleBreadcrumb from '../mixins/accessibleBreadcrumb'
import MixinFileActions from '../mixins/fileActions'

import { VisibilityObserver } from 'web-pkg/src/observer'
import { ImageDimension, ImageType } from '../constants'
import debounce from 'lodash-es/debounce'
import { bus } from 'web-pkg/src/instance'

import AppLoadingSpinner from 'web-pkg/src/components/AppLoadingSpinner.vue'
import NoContentMessage from 'web-pkg/src/components/NoContentMessage.vue'
import AppBar from '../components/AppBar/AppBar.vue'
import CreateAndUpload from '../components/AppBar/CreateAndUpload.vue'
import NotFoundMessage from '../components/FilesList/NotFoundMessage.vue'
import ListInfo from '../components/FilesList/ListInfo.vue'
import Pagination from '../components/FilesList/Pagination.vue'
import ContextActions from '../components/FilesList/ContextActions.vue'
import { breadcrumbsFromPath, concatBreadcrumbs } from '../helpers/breadcrumbs'
import { defineComponent } from '@vue/composition-api'
import { move } from '../helpers/resource'
import { Resource } from 'web-client'
import { usePublicLinkPassword, useStore } from 'web-pkg/src/composables'
import KeyboardActions from '../components/FilesList/KeyboardActions.vue'
import SideBar from '../components/SideBar/SideBar.vue'
import FilesViewWrapper from '../components/FilesViewWrapper.vue'

const visibilityObserver = new VisibilityObserver()

export default defineComponent({
  components: {
    FilesViewWrapper,
    AppBar,
    CreateAndUpload,
    ResourceTable,
    ListInfo,
    Pagination,
    AppLoadingSpinner,
    NoContentMessage,
    NotFoundMessage,
    ContextActions,
    KeyboardActions,
    SideBar
  },

  mixins: [MixinAccessibleBreadcrumb, MixinFileActions],

  setup() {
    const store = useStore()
    return {
      ...useResourcesViewDefaults<Resource, any, any[]>(),
      publicLinkPassword: usePublicLinkPassword({ store })
    }
  },

  computed: {
    ...mapGetters('Files', [
      'currentFolder',
      'highlightedFile',
      'totalFilesCount',
      'totalFilesSize'
    ]),
    ...mapGetters(['configuration']),

    breadcrumbs() {
      const breadcrumbs = breadcrumbsFromPath(this.$route, this.$route.params.item)
      const rootRoute = breadcrumbs.shift()

      return concatBreadcrumbs(
        { text: this.$gettext('Public link'), to: rootRoute.to },
        ...breadcrumbs
      )
    },

    isEmpty() {
      return this.paginatedResources.length < 1
    },

    folderNotFound() {
      return this.currentFolder === null
    },

    targetRoute() {
      return { name: this.$route.name }
    },

    displayThumbnails() {
      return !this.configuration?.options?.disablePreviews
    }
  },

  watch: {
    $route: {
      handler: function (to, from) {
        const sameRoute = to.name === from?.name
        const sameItem = to.params?.item === from?.params?.item
        if (!sameRoute || !sameItem) {
          this.loadResourcesTask.perform(this, sameRoute)
        }
      },
      immediate: true
    }
  },
  beforeDestroy() {
    visibilityObserver.disconnect()
  },

  created() {
    const loadResourcesEventToken = bus.subscribe('app.files.list.load', (path) => {
      this.loadResourcesTask.perform(this, this.$route.params.item === path, path)
    })

    this.$on('beforeDestroy', () => bus.unsubscribe('app.files.list.load', loadResourcesEventToken))
  },

  methods: {
    ...mapActions('Files', ['loadPreview']),
    ...mapMutations('Files', [
      'SET_CURRENT_FOLDER',
      'LOAD_FILES',
      'CLEAR_CURRENT_FILES_LIST',
      'REMOVE_FILES',
      'REMOVE_FILES_FROM_SEARCHED',
      'REMOVE_FILE_SELECTION'
    ]),

    async fileDropped(fileIdTarget) {
      const selected = [...this.selectedResources]
      const targetInfo = this.paginatedResources.find((e) => e.id === fileIdTarget)
      const isTargetSelected = selected.some((e) => e.id === fileIdTarget)
      if (isTargetSelected) return
      if (targetInfo.type !== 'folder') return
      const movedResources = await move(
        selected,
        targetInfo,
        this.$client,
        this.createModal,
        this.hideModal,
        this.showMessage,
        this.$gettext,
        this.$gettextInterpolate,
        this.$ngettext,
        false,
        this.publicLinkPassword
      )
      for (const resource of movedResources) {
        this.REMOVE_FILES([resource])
        this.REMOVE_FILES_FROM_SEARCHED([resource])
        this.REMOVE_FILE_SELECTION(resource)
      }
    },

    rowMounted(resource, component) {
      if (!this.displayThumbnails) {
        return
      }

      const debounced = debounce(({ unobserve }) => {
        unobserve()
        this.loadPreview({
          resource,
          isPublic: true,
          dimensions: ImageDimension.Thumbnail,
          type: ImageType.Thumbnail
        })
      }, 250)

      visibilityObserver.observe(component.$el, { onEnter: debounced, onExit: debounced.cancel })
    }
  }
})
</script>
