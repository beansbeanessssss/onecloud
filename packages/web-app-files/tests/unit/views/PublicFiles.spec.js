import { mount, RouterLinkStub } from '@vue/test-utils'
import { getStore, localVue, createFile, stubs } from './views.setup.js'
import FileActions from '@files/src/mixins/fileActions.js'
import MixinMountSideBar from '@files/src/mixins/sidebar/mountSideBar.js'
import PublicFiles from '@files/src/views/PublicFiles.vue'
import MixinFilesListPagination from '@files/src/mixins/filesListPagination'
import MixinAccessibleBreadcrumb from '@files/src/mixins/accessibleBreadcrumb'

const listLoaderStub = 'list-loader-stub'
const noContentStub = 'no-content-message-stub'
const filesTableStub = 'oc-table-files-stub'
const contextActionsStub = 'context-actions-stub'
const listInfoStub = 'list-info-stub'
const notFoundMessageStub = 'not-found-message-stub'
const selectors = {
  pagination: '.files-pagination'
}
const defaultRoute = { path: '/some-path', name: 'some-route-name', params: { page: 1 } }

localVue.prototype.$client.publicFiles = {
  list: async () => [
    {
      fileInfo: { '{http://owncloud.org/ns}fileid': 1234 },
      name: 'simple',
      getProperty: jest.fn()
    }
  ]
}

jest
  .spyOn(MixinFilesListPagination.methods, '$_filesListPagination_updateCurrentPage')
  .mockImplementation()

jest
  .spyOn(MixinAccessibleBreadcrumb.methods, 'accessibleBreadcrumb_focusAndAnnounceBreadcrumb')
  .mockImplementation()

const createFolder = ({ name = '1234', canCreate = false } = {}) => ({
  name: name,
  canCreate: () => canCreate
})

describe('PublicFiles', () => {
  describe('list loader component', () => {
    it('should be visible if loading is set as true', () => {
      const wrapper = getMountedWrapper({ loading: true })

      expect(wrapper.find(listLoaderStub).exists()).toBeTruthy()
    })

    it('should not be visible if loading is set as false', () => {
      const wrapper = getMountedWrapper({ loading: false })

      expect(wrapper.find(listLoaderStub).exists()).toBeTruthy()
    })
  })

  describe('not found message component', () => {
    stubs['not-found-message'] = true
    it('should be visible if current folder is null', async () => {
      const store = createStore({ currentFolder: null })
      const wrapper = getMountedWrapper({ store, loading: false })
      await wrapper.vm.$nextTick()

      expect(wrapper.find(notFoundMessageStub).exists()).toBeTruthy()
    })
    it('should not be visible if current folder is not null', async () => {
      const store = createStore({ currentFolder: createFolder() })
      const wrapper = getMountedWrapper({ store, loading: false })
      await wrapper.vm.$nextTick()

      expect(wrapper.find(notFoundMessageStub).exists()).toBeFalsy()
    })
  })

  describe('no content message component', () => {
    it('should be visible if active files length is less than one', async () => {
      stubs['no-content-message'] = false
      const store = createStore({ currentFolder: createFolder(), activeFiles: [] })
      const wrapper = getMountedWrapper({ store, loading: false })
      await wrapper.vm.$nextTick()

      expect(wrapper.find('#files-public-list-empty').exists()).toBeTruthy()
      expect(wrapper.find('#files-public-list-empty')).toMatchSnapshot()
      stubs['no-content-message'] = true
    })
    it('should not be visible if active files length is greater than one', async () => {
      const store = createStore({
        currentFolder: createFolder(),
        activeFiles: [createFile({ id: '1234' })]
      })
      const wrapper = getMountedWrapper({ store, loading: false })

      await wrapper.vm.$nextTick()

      expect(wrapper.find(noContentStub).exists()).toBeFalsy()
    })

    describe('call to action', () => {
      it('should show call to action if current folder has truthy "canCreate" value', async () => {
        stubs['no-content-message'] = false

        const store = createStore({
          currentFolder: createFolder({ canCreate: true }),
          activeFiles: []
        })
        const wrapper = getMountedWrapper({ store, loading: false })
        await wrapper.vm.$nextTick()

        const callToActionSpan = wrapper.find('[data-testid="public-files-call-to-action"]')

        expect(callToActionSpan.exists()).toBeTruthy()
        expect(callToActionSpan).toMatchSnapshot()
      })
      it('should show call to action if current folder has falsy "canCreate" value', async () => {
        stubs['no-content-message'] = false

        const store = createStore({
          currentFolder: createFolder({ canCreate: false }),
          activeFiles: []
        })
        const wrapper = getMountedWrapper({ store, loading: false })
        await wrapper.vm.$nextTick()

        const callToActionSpan = wrapper.find('[data-testid="public-files-call-to-action"]')

        expect(callToActionSpan.exists()).toBeFalsy()
      })
    })
  })

  describe('when length of active files is greater than zero', () => {
    const activeFiles = [createFile({ id: '1234' }), createFile({ id: '5896' })]
    const store = createStore({ activeFiles: activeFiles, currentFolder: createFolder() })
    const wrapper = getMountedWrapper({ store, loading: false })

    it('should load files table with correct props', () => {
      const filesTable = wrapper.find(filesTableStub)

      expect(filesTable.exists()).toBeTruthy()
      expect(filesTable.props()).toMatchObject({
        resources: activeFiles,
        arePathsDisplayed: false,
        areThumbnailsDisplayed: !store.getters.configuration.options.disablePreviews,
        areResourcesClickable: true,
        isSelectable: true,
        disabled: null,
        dragDrop: false,
        selection: [],
        hasActions: true,
        targetRoute: { name: defaultRoute.name }
      })
    })

    it('should show context menu for each resource', async () => {
      stubs['oc-table-files'] = false
      stubs['context-actions'] = true

      const wrapper = getMountedWrapper({ store, loading: false })
      await wrapper.vm.$nextTick()

      const contextActions = wrapper.findAll(contextActionsStub)

      expect(contextActions.length).toBe(activeFiles.length)
      activeFiles.forEach((activeFile, index) => {
        expect(contextActions.at(index).props()).toMatchObject({
          item: activeFile
        })
      })
    })

    it('should show list info component for oc files table', async () => {
      stubs['oc-table-files'] = false
      stubs['list-info'] = true

      const store = createStore({
        currentFolder: createFolder(),
        activeFiles: activeFiles,
        totalFilesCount: {
          files: 22,
          folders: 56
        },
        totalFileSize: 12545
      })

      const wrapper = getMountedWrapper({ store, loading: false })
      await wrapper.vm.$nextTick()
      const listInfo = wrapper.find(listInfoStub)

      expect(listInfo.exists()).toBeTruthy()
      expect(listInfo.props()).toMatchObject({
        files: store.getters['Files/totalFilesCount'].files,
        folders: store.getters['Files/totalFilesCount'].folders,
        size: store.getters['Files/totalFilesSize']
      })
    })

    describe('pagination component in oc files table', () => {
      it('should be visible if pages is greater than one', async () => {
        stubs['oc-table-files'] = false
        stubs.pagination = false
        stubs.RouterLink = RouterLinkStub
        const store = getStore({
          currentFolder: createFolder(),
          activeFiles: activeFiles,
          pages: 2,
          currentPage: 1
        })
        const wrapper = getMountedWrapper({ store, loading: false })
        await wrapper.vm.$nextTick()

        const pagination = wrapper.find(selectors.pagination)

        expect(pagination.exists()).toBeTruthy()
        stubs['oc-table-files'] = true
      })

      it('should not be visible if pages is less than one', async () => {
        stubs['oc-table-files'] = false
        stubs.pagination = false
        stubs.RouterLink = RouterLinkStub
        const store = getStore({
          currentFolder: createFolder(),
          activeFiles: activeFiles,
          pages: 1,
          currentPage: 1
        })
        const wrapper = getMountedWrapper({ store, loading: false })
        await wrapper.vm.$nextTick()

        const pagination = wrapper.find(selectors.pagination)

        expect(pagination.exists()).toBeFalsy()
        stubs['oc-table-files'] = true
      })
    })

    it('should call "$_mountSideBar_showDefaultPanel" method if "showDetails" event is emitted from oc files table component', async () => {
      const spyMountSideBarShowDefaultPanel = jest.spyOn(
        MixinMountSideBar.methods,
        '$_mountSideBar_showDefaultPanel'
      )
      const wrapper = getMountedWrapper({ store, loading: false })
      await wrapper.vm.$nextTick()

      const filesTable = wrapper.find(filesTableStub)
      expect(spyMountSideBarShowDefaultPanel).toHaveBeenCalledTimes(0)

      filesTable.vm.$emit('showDetails')

      expect(spyMountSideBarShowDefaultPanel).toHaveBeenCalledTimes(1)
    })

    it('should call "$_fileActions_triggerDefaultAction" method if "fileClick" event is emitted from oc files table component', async () => {
      const spyFileActionsTriggerDefaultAction = jest
        .spyOn(FileActions.methods, '$_fileActions_triggerDefaultAction')
        .mockImplementation()

      const wrapper = getMountedWrapper({ store, loading: false })
      await wrapper.vm.$nextTick()

      const filesTable = wrapper.find(filesTableStub)
      expect(spyFileActionsTriggerDefaultAction).toHaveBeenCalledTimes(0)

      filesTable.vm.$emit('fileClick')

      expect(spyFileActionsTriggerDefaultAction).toHaveBeenCalledTimes(1)
    })

    it('should call "rowMounted" method if "rowMounted" event is emitted from oc files table component', async () => {
      const spyRowMounted = jest.spyOn(PublicFiles.methods, 'rowMounted').mockImplementation()

      const wrapper = getMountedWrapper({ store, loading: false })
      await wrapper.vm.$nextTick()

      const filesTable = wrapper.find(filesTableStub)
      expect(spyRowMounted).toHaveBeenCalledTimes(0)

      filesTable.vm.$emit('rowMounted')

      expect(spyRowMounted).toHaveBeenCalledTimes(1)
    })
  })

  function createStore({
    activeFiles = [],
    sidebarClosed = true,
    disablePreviews = false,
    totalFilesCount,
    currentFolder = null,
    totalFileSize = null
  } = {}) {
    return getStore({
      activeFiles,
      sidebarClosed,
      disablePreviews,
      totalFilesCount,
      currentFolder,
      totalFileSize
    })
  }

  function getMountOptions({ $route = defaultRoute, store = createStore(), loading = true }) {
    return {
      localVue,
      store,
      stubs,
      mocks: {
        $route
      },
      data: () => ({
        loading: loading
      })
    }
  }

  function getMountedWrapper({ store, loading } = {}) {
    const component = { ...PublicFiles, created: jest.fn(), mounted: jest.fn() }
    return mount(component, getMountOptions({ store, loading }))
  }
})
