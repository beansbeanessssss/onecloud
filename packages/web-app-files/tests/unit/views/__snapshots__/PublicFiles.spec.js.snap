// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PublicFiles no content message component call to action should show call to action if current folder has truthy "canCreate" value 1`] = `
<span data-testid="public-files-call-to-action" data-msgid="
          Drag files and folders here or use the &quot;+ New&quot; button to upload
        " data-current-language="en_US">
          Drag files and folders here or use the "+ New" button to upload
        </span>
`;

exports[`PublicFiles no content message component should be visible if active files length is less than one 1`] = `
<div class="uk-height-1-1 uk-flex uk-flex-column uk-flex-center uk-flex-middle uk-text-center files-empty" id="files-public-list-empty">
  <div class="oc-icon oc-icon-xxl oc-icon-passive">
    <!---->
  </div>
  <div class="oc-text-muted uk-text-large"><span data-msgid="There are no resources in this folder" data-current-language="en_US">There are no resources in this folder</span></div>
  <div class="oc-text-muted"></div>
</div>
`;
